include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+    static this_t create(NBL_CONST_REF_ARG(LightSample) _sample, NBL_CONST_REF_ARG(Interaction) interaction, NBL_CONST_REF_ARG(Cache) cache, BxDFClampMode _clamp)
+    {
+        impl::__extract_aniso_vars<LightSample, Interaction, Scalar, interaction_traits<Interaction>::is_aniso> vars = impl::__extract_aniso_vars<LightSample, Interaction, Scalar, interaction_traits<Interaction>::is_aniso>::create(_sample, interaction);
+        impl::__extract_aniso_vars2<Cache, Scalar, interaction_traits<Interaction>::is_aniso> vars2 = impl::__extract_aniso_vars2<Cache, Scalar, interaction_traits<Interaction>::is_aniso>::create(cache);
+
+        this_t retval;
+        retval.NdotH = vars2.NdotH;
+        retval.NdotH2 = vars2.NdotH2;
+        retval.NdotV = math::conditionalAbsOrMax<Scalar>(_clamp == BxDFClampMode::BCM_ABS, vars.NdotV, 0.0);
+        retval.uNdotV = vars.NdotV;
+        retval.NdotV2 = vars.NdotV2;
+        retval.uNdotL = _sample.getNdotL();
+        retval.NdotL = math::conditionalAbsOrMax<Scalar>(_clamp == BxDFClampMode::BCM_ABS, retval.uNdotL, 0.0);
+        retval.NdotL2 = _sample.getNdotL2();
+        retval.VdotL = _sample.getVdotL();
+        retval.VdotH = vars2.VdotH;
+        retval.LdotH = vars2.LdotH;
+
+        retval.is_aniso = interaction_traits<Interaction>::is_aniso;
+        retval.TdotL2 = vars.TdotL2;
+        retval.BdotL2 = vars.BdotL2;
+        retval.TdotV2 = vars.TdotV2;
+        retval.BdotV2 = vars.BdotV2;
+        retval.TdotH2 = vars2.TdotH2;
+        retval.BdotH2 = vars2.BdotH2;
+        return retval;
+    }
+
+    Scalar getMaxNdotV() { return max<Scalar>(uNdotV, 0.0); }
+    Scalar getAbsNdotV() { return abs<Scalar>(uNdotV); }
+
+    Scalar getMaxNdotL() { return max<Scalar>(uNdotL, 0.0); }
+    Scalar getAbsNdotL() { return abs<Scalar>(uNdotL); }
+
+    // iso
+    Scalar NdotH;
+    Scalar NdotH2;
+    Scalar NdotV;
+    Scalar NdotV2;
+    Scalar NdotL;
+    Scalar NdotL2;
+    Scalar VdotH;
+    Scalar LdotH;
+    Scalar VdotL;
+
+    // aniso
+    bool is_aniso;
+    Scalar TdotH2;
+    Scalar BdotH2;
+    Scalar TdotL2;
+    Scalar BdotL2;
+    Scalar TdotV2;
+    Scalar BdotV2;
+
+    // original, unclamped
+    Scalar uNdotL;
+    Scalar uNdotV;
+};
+
+// unified param struct for calls to BxDF::create
+template<typename Scalar, typename Spectrum NBL_PRIMARY_REQUIRES(is_scalar_v<Scalar>)
+struct SBxDFCreationParams
+{
+    bool is_aniso;
+    vector<Scalar, 2> A;    // roughness
+    Spectrum ior0;          // source ior
+    Spectrum ior1;          // destination ior
+    Scalar eta;             // in most cases, eta will be calculated from ior0 and ior1; see monochromeEta in pathtracer.hlsl
+    Spectrum eta2;
+    Spectrum luminosityContributionHint;
+};
```

Comment:
we don't want/need this

Reply:
This is something I added just for use in the path tracer example. Because the BxDF materials there are in an array and only consist of the creation params and sometimes they overlap (a `float3` for both real ior and albedo depending on diffuse vs conductor). So I chose to add this and each bxdf can handle it differently. Though it really doesn't look good.

Alternatively, I'd move this to the path tracer example only and add another set of switch statements to handle which params go to which bxdf type (diffuse, conductor, dielectric).

Reply:
move it to path tracer example, you're allowed to do whatever there

Reply:
continued in https://github.com/Devsh-Graphics-Programming/Nabla/pull/899/files?diff=unified&w=1#r2185169394